src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp

```diff
@@ -941,6 +945,25 @@ namespace Opm {
         }
     }
 
+    void Schedule::handleWFOAM( const DeckKeyword& keyword, size_t currentStep, const ParseContext& parseContext, ErrorGuard& errors) {
+        for (const auto& record : keyword) {
+            const std::string& wellNamePattern = record.getItem("WELL").getTrimmedString(0);
+            const auto well_names = wellNames(wellNamePattern, currentStep );
+
+            if (well_names.empty())
+                invalidNamePattern(wellNamePattern, parseContext, errors, keyword);
+
+            for (const auto& well_name : well_names) {
+                auto& dynamic_state = this->wells_static.at(well_name);
```

Comment:
While I think I understand the general idea here (call updateWell with a new well object if changes are required due to a WFOAM keyword) there are some unclear things:
 - Why do we use a mutable reference to `dynamic_state`?
 - Is there some sharing of data between what is already in the `dynamic_state` and what gets into `well2` or is well2 completely independent (i.e. is a copy of a `Well2` a deep copy)?

Reply:
>  Why do we use a mutable reference to dynamic_state

That is a mistake - it should have been `const`. The lines of code you are looking at right now do not mutate the `dynamic_state` instance - if there is a net change the `dynamic_state` will be updated in the `updateWell()` method.

> Is there some sharing of data between what is already in the dynamic_state

There should be no sharing of state, the copy constructor of the `Well2` is invoked - and as long as that is deep (which it is) everything is deep. Just looking at the code here - you could in principle be fooled by a shallow copy constructor - that I agree. The update goes like this:

1. Take a deep copy.
2. Modify the copy with e.g. updated foam properties.
3. If step 2 gives rise to a net change in the well object: install the new well object in the dynamic state, otherwise it is discarded again.